### PR TITLE
Issue 4874: Upgrade HDFS Client from 2.x to 3.x in order to address security vulnerabilities

### LINKED ISSUE #4874: Fix HDFS security vulnerabilities
**Problem description**
Pravega uses HDFS for tier 2 and it has a set of security vulnerabilities. These include `CVE-2016-6811, CVE-2018-8009, CVE-2018-8029, CVE-2018-11766`. These come from the Apache Hadoop hdfs library along with its transitive dependencies. The versions of these libraries needs to be updated to overcome the security vulnerabilities. The updates to the library requires changes in Pravega source code to remove the methods, exceptions and properties used from the library that are no longer available and to use the replacements for these and other deprecated syntax.

**Problem location**
The version for the Apache hadoop hdfs library is specified in the gradle.properties. Its usage spans HDFS tests and exception helpers. These changes expand to others parts of the code where they are used. 

**Suggestions for an improvement**
 The version 2.9.2 of apache hadoop hdfs fixes the above CVEs but there are a few more CVEs that have been addressed in recent versions. These new versions have increased the breaks in compatibility with our existing code. The updates can therefore be progressive increasing from 2.9.2 to more recent versions. The changes to the hdfs library and its transitive dependencies can be made at once.

### PR DESCRIPTION
**Pull request by ravisharda**
**Change log description**  
Upgrade HDFS client version to the latest version, which is currently `3.3.0` as per [Maven Repo](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-hdfs).  

**Purpose of the change**  
Fixes #4874. 

**What the code does**  

* Updates HDFS client dependency. 
* Modifies code to account for breaking changes. 
* Adds additional dependencies to `bindings` project (since the newer version doesn't brings some of the transitive dependencies that were used by other components of the project)

**How to verify it**  
All unit and integration tests must pass. 

In addition, a distributed mode cluster running HDFS, must be able to run successfullly and services reads and writes. I use Docker-Compose based cluster for this purpose and deployed it using this YAML: 
https://github.com/ravisharda/pravega/blob/issue-4874-upgrade-hdfs/docker/compose/docker-compose.yml

**Note:** 
1. Note that the YAML runs `pravega/hdfs:2.7.7` as the HDFS server, which is:
   * Built out of code here: https://github.com/pravega/docker-hdfs
   * Based on Hadoop v2.7.7

   So, the test mentioned above uses HDFS client `v3.3.0` against a server `v2.7.7`. 

   Upgrade of the server should be taken separately as part of a different PR for the https://github.com/pravega/docker-hdfs repo. 

2. This PR borrows some of the work done in an older and now closed PR #4878 by @ravibeta.

### CODE REVIEW

**Review by sachin-j-joshi — APPROVED** [review]
Looks okay to me.

**Review by sachin-j-joshi — COMMENTED** [review]
Are there more changes expected ?

**Comment by ravisharda** [comment]
> Are there more changes expected ?

@sachin-j-joshi Yes, and I've now made those changes with the last commit.

File: bindings/src/test/java/io/pravega/storage/extendeds3/S3ProxyImpl.java [hunk 1]
```diff
@@ -37,7 +37,7 @@
 import lombok.val;
 import org.apache.commons.httpclient.HttpStatus;
 import org.apache.commons.io.IOUtils;
-import org.apache.commons.lang.ArrayUtils;
+import org.apache.commons.lang3.ArrayUtils;
```
Comment (andreipaduroiu) [inline bindings/src/test/java/io/pravega/storage/extendeds3/S3ProxyImpl.java:40]: Unrelated.

Comment (ravisharda) [reply bindings/src/test/java/io/pravega/storage/extendeds3/S3ProxyImpl.java:40]: This was necessary, as it was depending on a transitive dependency coming via Hadoop client. Once I upgraded Hadoop client, the code compilation failed here. So, I used the replacement that was being brought in.

**Review by RaulGracia — APPROVED** [review]
LGTM, but it would be advisable to run system tests with HDFS (docker swarm) against this branch.

**Comment by ravisharda** [comment]
> LGTM, but it would be advisable to run system tests with HDFS (docker swarm) against this branch.

@RaulGracia I've run this branch successfully with HDFS using a Docker Compose cluster instead (as described in the `Description` > `How to Test` section).

**Review by andreipaduroiu — APPROVED** [review]

**Pull request merged**